buildreport/buildreport.go

```diff
@@ -0,0 +1,454 @@
+				return err
+			}
+			return gitcmd.Run(gitDir, "push", auther.InsertAuth(url), "HEAD:"+githubWikiDefaultBranch)
+		})
+		if err != nil {
+			// Inner retry wasn't able to get the update done. This may be due to concurrency: N
+			// builds trying to update the issue at the same time. Try again after a short delay.
+			// (Nothing fancy: even in the worst case of N updates happening simultaneously,
+			// repeatedly, all concurrent updates will eventually be able to get through because one
+			// update out of N succeeds each time.)
+			log.Printf(
+				"Inner GitHub retry loop failed. Waiting %v then trying again. Will give up after %v. Error: %v\n",
+				reportRetryDelayDuration,
+				reportRetryDelayDuration-elapsed,
+				err)
+			time.Sleep(reportRetryDelayDuration)
+			continue
+		}
+		break // Success.
+	}
+	// Now that we've successfully pushed, the data is saved, and we know it's the latest
+	// available data. Update the issue.
+	//
+	// There is potential for a conflict with another simultaneous UpdateIssueBody call here: if
+	// there is a delay between "push" and Edit, and another UpdateIssueBody call starts and
+	// finishes during that delay, this Edit will revert the issue to show old data!
+	//
+	// Mitigations: there isn't any known reason we'd have a long delay right here. The data
+	// itself is safe: the next UpdateIssueBody call will fix the issue and post the correct
+	// data. The report includes a link to the actual data in case it's important to get the
+	// real data during a release.
+	log.Printf("Copying report to https://github.com/%v/%v/issues/%v description...", owner, repoName, issue)
+	return githubutil.Retry(func() error {
+		edit, _, err := client.Issues.Edit(ctx, owner, repoName, issue, &github.IssueRequest{Body: &body})
+		if err != nil {
+			return err
+		}
+		log.Printf("Edit successful:\n%v\n", edit.ID)
+		return nil
+	})
+}
+
+// Notify determines if a notification is necessary for the given status update and sends it.
+func Notify(ctx context.Context, owner string, repoName string, pat string, issue int, s State) error {
+	var notification string
+	switch s.Status {
+	case SymbolFailed:
+		notification = "Build failed!\n"
+	case SymbolSucceeded:
+		switch s.Name {
+		case releaseBuildPipelineName:
+			notification = "Completed releasing one version! If every version's release-build has finished, approve the release-go-images build to continue.\n"
+		case releaseImagesPipelineName:
+			notification = "Completed building all images! " +
+				"Before you [announce](https://github.com/microsoft/go-infra/blob/main/docs/release-process/instructions.md#making-the-internal-announcement), " +
+				"confirm the MAR/MCR images are updated using commands like these:\n\n" +
+				"```\n" +
+				"image=mcr.microsoft.com/oss/go/microsoft/golang:1-bullseye\n" +
+				"docker pull $image\n" +
+				"docker run -it --rm $image go version\n" +
+				"```\n"
+		}
+	}
+
+	if notification != "" {
```

Comment:
nit: if you negate this condition then you can return early and reduce one indentation level.

Reply:
Good spot, it also lets me return the error value from `githubutil.Retry` directly.